src/librustc_infer/traits/select.rs

```diff
@@ -1309,31 +1295,35 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             return;
         }
 
-        if self.can_use_global_caches(param_env) {
-            if let Err(Overflow) = candidate {
-                // Don't cache overflow globally; we only produce this in certain modes.
-            } else if !trait_ref.has_local_value() {
-                if !candidate.has_local_value() {
-                    debug!(
-                        "insert_candidate_cache(trait_ref={:?}, candidate={:?}) global",
-                        trait_ref, candidate,
-                    );
-                    // This may overwrite the cache with the same value.
-                    tcx.selection_cache
-                        .hashmap
-                        .borrow_mut()
-                        .insert(param_env.and(trait_ref), WithDepNode::new(dep_node, candidate));
-                    return;
-                }
-            }
+        // HACK(eddyb) never cache overflow (this check used to be global-only).
+        if let Err(Overflow) = candidate {
```

Comment:
with the exception of this tiny logic change, I guess

Reply:
I can't really think why we would want to cache overflow *locally* anyhow, that seems like a bug...

Reply:
Pfft, that's nothing, here's something more fun: I don't think we ever use the local caches except for the `intercrate` mode, because we're doing freshening.

So we could put `intercrate` in `Reveal` or w/e then replace freshening with canonicalization, and have only global caching (either that or I'm seriously misunderstanding the implementation).